### PR TITLE
GNU autotools build system

### PR DESCRIPTION
**Pull request by darealshinji**
Native Linux builds and cross-building with mingw32 works. Building tarballs with `make dist` and out-of-tree builds are working too.
I hope you will find this useful.

### CODE REVIEW

**Comment by Venti-** [comment]
Thanks. We will look into replacing the current makefile with this thing. Might take a while though, as we have to learn autotools a bit first.

I got as far as running "autoreconf -i" and compiled successfully after some fixes, but will have to look into it further to see what is different. Autotools seems pretty complicated, but so is the current makefile.

File: configure.ac [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                ASFLAGS="$ASFLAGS -fwin32 -DPREFIX -DHAVE_ALIGNED_STACK=0"
+               ], [
+                ASFLAGS="$ASFLAGS -fwin64 -DHAVE_ALIGNED_STACK=1"
+               ]
+         )
+        ], [
+         ASFLAGS="$ASFLAGS -f elf$BITS"
+         CFLAGS="$CFLAGS -pthread"
+         LDFLAGS="$LDFLAGS -pthread -Wl,-z,noexecstack"
+         LIBS="$LIBS -lpthread"
```
Comment (aryla) [inline configure.ac]: It looks like `-lrt` should be included here. Otherwise the build fails with "undefined reference to `clock_gettime'."

File: src/kvazaar.h [hunk 2]
```diff
@@ -34,6 +34,10 @@
 extern "C" {
 #endif
 
+#if defined(KVZ_STATIC_LIB) && defined(PIC)
+  #undef KVZ_STATIC_LIB
```
Comment (aryla) [inline src/kvazaar.h]: Why is this needed? It causes linking against a static kvazaar lib to fail if the user has defined `PIC`.

**Comment by aryla** [comment]
Definitely better than the current system. We still need to add some way to compile and run tests before we can scrap the old Makefile.

File: configure.ac (hunk 1 above)
Comment (darealshinji) [inline configure.ac]: Which OS/compiler? Somehow I haven't had any problems with Mingw32 and native gcc on Linux so far.

File: src/kvazaar.h (hunk 2 above)
Comment (darealshinji) [inline src/kvazaar.h]: This was a dirty workaround to pass separate flags for shared and static libraries, which is otherwise a bit difficult using libtool and automake. Maybe libtool should only be used to generate shared libraries and I should add a separate build target for the static library?

**Comment by darealshinji** [comment]
I've added tests. Now I'll think about a way to only build the static library with `-DKVZ_STATIC_LIB` but without entirely dropping libtool.

File: src/kvazaar.h (hunk 2 above)
Comment (darealshinji) [inline src/kvazaar.h]: Apparently gcc, clang and mingw32 require `KVZ_PUBLIC` to be defined as `__attribute__ ((visibility ("default")))` for dll builds (not required for static builds but it doesn't seem to do any harm either):
``` C
#if defined(__GNUC__)
  // Using GCC.
  #define KVZ_PUBLIC __attribute__ ((visibility ("default")))
#elif defined(_WIN32)
  // Using windows and not GCC.
  #if defined(KVZ_STATIC_LIB)
    // Using or building kvazaar as a static library on windows.
    #define KVZ_PUBLIC
  #elif defined(KVZ_DLL_EXPORTS) || defined(DLL_EXPORT)
    // Building kvazaar on windows.
    #define KVZ_PUBLIC __declspec(dllexport)
  #else
    // Using kvazaar as a DLL on windows.
    #define KVZ_PUBLIC __declspec(dllimport)
  #endif
#else
  #define KVZ_PUBLIC
#endif
```
This way `KVZ_DLL_EXPORTS` and `KVZ_STATIC_LIB` don't need to be defined in configure.ac or Makefile.am. But I'm not sure how handle non-gcc compatible compilers invoked through configure (though that might be relatively unlikely?).

File: configure.ac (hunk 1 above)
Comment (aryla) [inline configure.ac]: This happened with each of gcc 4.6.4, gcc 4.9.2 and clang 3.0 on Ubuntu 12.04 x86_64.

File: src/kvazaar.h (hunk 2 above)
Comment (aryla) [inline src/kvazaar.h]: I think we can use `defined(PIC)` to differentiate between shared and static builds as long as we check `defined(KVZ_DLL_EXPORTS)` to make sure that `PIC` is meant for kvazaar. Maybe something like this:
```c
#if defined(KVZ_DLL_EXPORTS)
  #if !defined(PIC)
    // Building static kvazaar library.
    #define KVZ_PUBLIC
  #elif defined(_WIN32) || defined(__CYGWIN__)
    // Building kvazaar DLL on Windows.
    #define KVZ_PUBLIC __declspec(dllexport)
  #elif defined(__GNUC__)
    // Building kvazaar shared library with GCC.
    #define KVZ_PUBLIC __attribute__ ((visibility ("default")))
  #else
    #define KVZ_PUBLIC
  #endif
#else
  #if defined(KVZ_STATIC_LIB)
    // Using static kvazaar library.
    #define KVZ_PUBLIC
  #elif defined(_WIN32) || defined(__CYGWIN__)
    // Using kvazaar DLL on Windows.
    #define KVZ_PUBLIC __declspec(dllimport)
  #else
    // Using kvazaar shared library and not on Windows.
    #define KVZ_PUBLIC
  #endif
#endif
```
